File: app_concierge/lib/services/cliente/client_company_service.dart
import 'package:app_concierge/features/data/domain/user_login_sqlite_service.dart';
import 'package:app_concierge/features/domain/client/client_company.dart';
import 'package:dio/dio.dart';
import 'package:app_concierge/core/constants/url_constants.dart';

class ClientCompanyService {
  Future<List<ClientCompany>> filterCode(String code) async {
    final dio = Dio();

    try {
      final response = await dio.get("$kURL_CLIENT_FILTER/code/$code");

      List<ClientCompany> list = [];
      for (var client in response.data) {
        list.add(ClientCompany.fromJson(client));
      }
      return list;
    } on DioException catch (e) {
      print('Erro: $e');
    }
    return List.empty();
  }

  Future<List<ClientCompany>> filterJFantasia(String fantasia) async {
    final dio = Dio();

    try {
      final response =
          await dio.get("$kURL_CLIENT_FILTER/j/fantasia/$fantasia");

      List<ClientCompany> list = [];
      for (var client in response.data) {
        list.add(ClientCompany.fromJson(client));
      }

      return list;
    } on DioException catch (e) {
      print('Erro: $e');
    }
    return List.empty();
  }

  Future<List<ClientCompany>> filterFFantasia(String fantasia) async {
    final dio = Dio();

    try {
      final response =
          await dio.get("$kURL_CLIENT_FILTER/f/fantasia/$fantasia");

      List<ClientCompany> list = [];
      for (var client in response.data) {
        list.add(ClientCompany.fromJson(client));
      }

      return list;
    } on DioException catch (e) {
      print('Erro: $e');
    }
    return List.empty();
  }

  Future<List<ClientCompany>> filterJName(String name) async {
    final dio = Dio();

    try {
      final response = await dio.get("$kURL_CLIENT_FILTER/j/name/$name");

      List<ClientCompany> list = [];
      for (var client in response.data) {
        list.add(ClientCompany.fromJson(client));
      }

      return list;
    } on DioException catch (e) {
      print('Erro: $e');
    }
    return List.empty();
  }

  Future<List<ClientCompany>> filterFName(String name) async {
    final dio = Dio();

    try {
      final response = await dio.get("$kURL_CLIENT_FILTER/f/name/$name");

      List<ClientCompany> list = [];
      for (var client in response.data) {
        list.add(ClientCompany.fromJson(client));
      }

      return list;
    } on DioException catch (e) {
      print('Erro: $e');
    }
    return List.empty();
  }

  Future<String> save(ClientCompany client) async {
    final dio = Dio();

    try {
      final token = await userStorange();

      final response = await dio.post(
        kURL_CLIENTCOMPANY_SAVE,
        data: client.toJson(),
        options: Options(headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $token'
        }, responseType: ResponseType.json),
      );
      if (response.statusCode == 201) {
        return "Success.";
      }
      return "Error.";
    } on DioException catch (e) {
      return "Error.";
    }
  }

  Future<String> userStorange() async {
    UserLoginSqliteService service = UserLoginSqliteService();
    var login = await service.userLogin();
    return login.token!;
  }
}

File: system-concierge/src/app/services/clientecompany/client-company.service.ts
import { Injectable } from '@angular/core';
import { environment } from '../../../environments/environment';
import { HttpClient, HttpHeaders, HttpResponse } from '@angular/common/http';
import { Observable } from 'rxjs';
import { StorageService } from '../storage/storage.service';
import { ClientCompany } from '../../models/clientcompany/client-company';
import { MessageResponse } from '../../models/message/message-response';

@Injectable({
  providedIn: 'root'
})
export class ClientCompanyService {

  companyResale: string = this.storage.companyId + "/" + this.storage.resaleId;

  constructor(private http: HttpClient, private storage: StorageService) { }

  save(client: ClientCompany): Observable<HttpResponse<MessageResponse>> {
    return this.http.post<MessageResponse>(environment.apiuUrl + "/clientcompany/save", client, { headers: this.myHeaders(), observe: 'response' });
  }
  update(client: ClientCompany): Observable<HttpResponse<MessageResponse>> {
    return this.http.post<MessageResponse>(environment.apiuUrl + "/clientcompany/update", client, { headers: this.myHeaders(), observe: 'response' });
  }
  listAll(): Observable<HttpResponse<MessageResponse>> {
    return this.http.get<MessageResponse>(environment.apiuUrl + "/clientcompany/" + this.companyResale + "/filter/all", { headers: this.myHeaders(), observe: 'response' });
  }
  filterId(id: Number): Observable<HttpResponse<MessageResponse>> {
    return this.http.get<MessageResponse>(environment.apiuUrl + "/clientcompany/" + this.companyResale + "/filter/id/" + id, { headers: this.myHeaders(), observe: 'response' });
  }
  filterJFantasia(name: string): Observable<HttpResponse<MessageResponse>> {
    return this.http.get<MessageResponse>(environment.apiuUrl + "/clientcompany/" + this.companyResale + "/filter/j/fantasia/" + name, { headers: this.myHeaders(), observe: 'response' });
  }
  filterFFantasia(name: string): Observable<HttpResponse<MessageResponse>> {
    return this.http.get<MessageResponse>(environment.apiuUrl + "/clientcompany/" + this.companyResale + "/filter/f/fantasia/" + name, { headers: this.myHeaders(), observe: 'response' });
  }
  filterJName(name: string): Observable<HttpResponse<MessageResponse>> {
    return this.http.get<MessageResponse>(environment.apiuUrl + "/clientcompany/" + this.companyResale + "/filter/j/name/" + name, { headers: this.myHeaders(), observe: 'response' });
  }
  filterFName(name: string): Observable<HttpResponse<MessageResponse>> {
    return this.http.get<MessageResponse>(environment.apiuUrl + "/clientcompany/" + this.companyResale + "/filter/f/name/" + name, { headers: this.myHeaders(), observe: 'response' });
  }
  filterCNPJ(cnpj: string): Observable<HttpResponse<MessageResponse>> {
    return this.http.get<MessageResponse>(environment.apiuUrl + "/clientcompany/" + this.companyResale + "/filter/cnpj/" + cnpj, { headers: this.myHeaders(), observe: 'response' });
  }
  filterCPF(cpf: string): Observable<HttpResponse<MessageResponse>> {
    return this.http.get<MessageResponse>(environment.apiuUrl + "/clientcompany/" + this.companyResale + "/filter/cpf/" + cpf, { headers: this.myHeaders(), observe: 'response' });
  }
  
  private myHeaders(): HttpHeaders {
    const httpOptions = new HttpHeaders({
      'Content-Type': 'application/json',
      'Authorization': 'Bearer ' + this.storage.token,
    });
    return httpOptions;
  }
}
